ENT-3847: Added serialization functions for sequence of strings

olehermanse opened the pull request.

olehermanse commented:
@cf-bottom please trigger a jenkins build for this as well

cf-bottom commented:
Alright, I triggered a build:

[![Build Status](https://ci.cfengine.com/buildStatus/icon?job=pr-pipeline&build=746)](https://ci.cfengine.com/job/pr-pipeline/746/)

https://ci.cfengine.com/job/pr-pipeline/746/

vpodzime commented on libutils/sequence.c:
```diff
@@ -449,3 +450,87 @@ int SeqStringLength(Seq *seq)
+// TODO: These static helper functions could be (re)moved
+static bool HasNulByte(const char *str, size_t n)
+{
+    for (int i = 0; i<n; ++i)
```
missing space around ``<``

karlhto replied:
+1

vpodzime commented on libutils/sequence.c:
```diff
@@ -449,3 +450,87 @@ int SeqStringLength(Seq *seq)
+        return -1;
+    }
+    size_t length;
+    int ret = sscanf(data, "%10zd", &length);
```
What is the point of reading ``size_t`` here and then converting it to ``long``?

vpodzime replied:
Also, this should use ``PREFIX_LENGTH``.

olehermanse replied:
The length is a size_t, the return value is a long (since it can be -1).

vpodzime replied:
I believe ``size_t`` can also be -1.

vpodzime commented on libutils/sequence.c:
```diff
@@ -449,3 +450,87 @@ int SeqStringLength(Seq *seq)
+    return (long)length;
+}
+
+char *SeqStringSerialize(Seq *seq)
```
Do we also need a function to write into a file? The resulting string may potentially be quite big.

olehermanse replied:
Yes, definitely.

vpodzime commented on libutils/sequence.c:
```diff
@@ -449,3 +450,87 @@ int SeqStringLength(Seq *seq)
+    for (int i = 0; i < length; ++i)
+    {
+        const char *s = SeqAt(seq, i);
+        WriterWriteF(w, "%-9ld %s\n", strlen(s), s);
```
Should use ``PREFIX_LENGTH``.

vpodzime commented on libutils/sequence.c:
```diff
@@ -449,3 +450,87 @@ int SeqStringLength(Seq *seq)
+
+        // Copy, NUL-terminate and add to Sequence
+        memcpy(new_str, src, length);
+        new_str[length] = '\0';
```
I think you could use ``strndup()`` instead of the ``xmalloc(); memcpy(); new_str[length] = '\0'`` here.

olehermanse commented on libutils/sequence.c:
```diff
@@ -368,7 +369,33 @@ Seq *SeqGetRange(const Seq *seq, size_t start, size_t end)
+void SeqRemoveNulls(Seq *s)
+{
+    int length = SeqLength(s);
+    int from = 0;
```
I didn't change this code. just move it. And we don't have any strict rules, asserts are "optional".

olehermanse commented:
@vpodzime Please re-review. I have addressed your comments, and added a StringCopy function so the deserialization can be done in one pass. I would like to add the file stream based versions in a separate PR.

olehermanse commented:
@cf-bottom please trigger a new jenkins build for this PR.

cf-bottom commented:
Alright, I triggered a build:

[![Build Status](https://ci.cfengine.com/buildStatus/icon?job=pr-pipeline&build=759)](https://ci.cfengine.com/job/pr-pipeline/759/)

https://ci.cfengine.com/job/pr-pipeline/759/

olehermanse commented:
@vpodzime Please re-review. I have addressed your comments, and added a StringCopy function so the deserialization can be done in one pass. I would like to add the file stream based versions in a separate PR.

olehermanse commented:
@cf-bottom test

cf-bottom commented:
@olehermanse : I'm sorry, I cannot do that. @olehermanse please help.

olehermanse commented:
@cf-bottom test

cf-bottom commented:
@olehermanse : I'm sorry, I cannot do that. @olehermanse please help.

cf-bottom commented:
@olehermanse : I'm sorry, I cannot do that. @olehermanse please help.

cf-bottom commented:
I'm not sure I understand, @olehermanse.

oleorhagen commented:
@cf-bottom please trigger a build

cf-bottom commented:
@oleorhagen : I'm sorry, I cannot do that. @olehermanse please help.

oleorhagen commented:
@olehermanse, please help!

olehermanse commented:
@vpodzime @karlhto Please review :)

karlhto commented on libutils/sequence.h:
```diff
@@ -53,6 +53,8 @@ typedef struct
 
 static inline void *SeqAt(const Seq *seq, int i)
 {
+    assert(seq != NULL);
```
Nice! Thought about this when I looked at the code earlier.

karlhto commented on libutils/sequence.c:
```diff
@@ -449,3 +450,103 @@ int SeqStringLength(Seq *seq)
+#define SEQ_PREFIX_LEN 10
+
+// TODO: These static helper functions could be (re)moved
+static bool HasNulByte(const char *str, size_t n)
```
Missing an L?

olehermanse replied:
Haha, @vpodzime ! This is exactly what I said to him, but apparently calling it `NUL` byte and `NULL` pointer is popular (in some circles, I guess).

karlhto replied:
Pretty sure it's called NUL-byte because the ASCII-"names" are limited to three characters.

karlhto replied:
The actual name should probably be HasNullByte still :P

olehermanse replied:
@karlhto Not necessarily. If enough people use the term Nul-byte (with one l) it makes sense to call it that. I'm not sure though.

karlhto replied:
We can compromise and you can call it HasNullCharacter

karlhto commented on libutils/sequence.c:
```diff
@@ -449,3 +450,103 @@ int SeqStringLength(Seq *seq)
+    for (int i = 0; i < length; ++i)
+    {
+        const char *s = SeqAt(seq, i);
+        WriterWriteF(w, "%-" STR_EXPAND(SEQ_PREFIX_LEN) "ld%s\n", strlen(s), s);
```
Maybe split up this line

karlhto commented on libutils/sequence.c:
```diff
@@ -449,3 +450,103 @@ int SeqStringLength(Seq *seq)
+    {
+        return -1;
+    }
+    if (!isdigit(data[0]))
```
Could have more space between if-statements for clarity

karlhto replied:
Same for subsequent statements

olehermanse commented:
@cf-bottom Please trigger another build

olehermanse commented:
@cf-bottom jenkins!

cf-bottom commented:
Sure, I triggered a build:

[![Build Status](https://ci.cfengine.com/buildStatus/icon?job=pr-pipeline&build=893)](https://ci.cfengine.com/job/pr-pipeline/893/)

https://ci.cfengine.com/job/pr-pipeline/893/

vpodzime commented:
One simply **has** to love C!
```
20:25:26 In file included from stpncpy.c:24:0:
20:25:26 stpncpy.c: In function 'stpncpy':
20:25:26 stpncpy.c:40:19: error: 'NULL' undeclared (first use in this function)
20:25:26      assert(dst != NULL);
20:25:26                    ^
20:25:26 stpncpy.c:40:19: note: each undeclared identifier is reported only once for each function it appears in
20:25:26 Makefile:562: recipe for target 'stpncpy.lo' failed
```

olehermanse commented:
@cf-bottom jenkins!

cf-bottom commented:
Sure, I triggered a build:

[![Build Status](https://ci.cfengine.com/buildStatus/icon?job=pr-pipeline&build=904)](https://ci.cfengine.com/job/pr-pipeline/904/)

https://ci.cfengine.com/job/pr-pipeline/904/

vpodzime commented:
A build including exotics:
[![Build Status](https://ci.cfengine.com/buildStatus/icon?job=pr-pipeline&build=907)](https://ci.cfengine.com/job/pr-pipeline/907/)

olehermanse commented:
@cf-bottom jenkins test

olehermanse commented:
@cf-bottom jenkins test

cf-bottom commented:
Alright, I triggered a build:

[![Build Status](https://ci.cfengine.com/buildStatus/icon?job=pr-pipeline&build=926)](https://ci.cfengine.com/job/pr-pipeline/926/)

https://ci.cfengine.com/job/pr-pipeline/926/

olehermanse commented:
@cf-bottom jenkins test

cf-bottom commented:
Sure, I triggered a build:

[![Build Status](https://ci.cfengine.com//buildStatus/icon?job=pr-pipeline&build=928)](https://ci.cfengine.com//job/pr-pipeline/928/)

https://ci.cfengine.com/job/pr-pipeline/928/

olehermanse commented:
This is the info from jenkins test:

```
12:44:13 seq != NULL
12:44:13 ERROR: sequence_test.c:437 Failure!
12:44:13 seq2 != NULL
12:44:13 ERROR: sequence_test.c:486 Failure!
12:44:13 ==================================================
12:44:13 Starting test: sequence_test.c
```

olehermanse commented:
xcalloc: 
[![Build Status](https://ci.cfengine.com/buildStatus/icon?job=pr-pipeline&build=929)](https://ci.cfengine.com/job/pr-pipeline/929/)

olehermanse commented:
strncpy:
[![Build Status](https://ci.cfengine.com/buildStatus/icon?job=pr-pipeline&build=930)](https://ci.cfengine.com/job/pr-pipeline/930/)

olehermanse commented:
Looks like this was fixed by `calloc`. I'll have a closer look tomorrow.

olehermanse commented:
Ran the strncpy version that passed, on all exotics:

[![Build Status](https://ci.cfengine.com/buildStatus/icon?job=pr-pipeline&build=938)](https://ci.cfengine.com/job/pr-pipeline/938/)

Ran the version using fallback on all platforms:

[![Build Status](https://ci.cfengine.com/buildStatus/icon?job=pr-pipeline&build=940)](https://ci.cfengine.com/job/pr-pipeline/940/)

olehermanse commented:
This pull request **introduces 1 alert** when merging 2bd872814fe936d6ab63744fe7396d9fbacbac3c into bfe82403979f7780b06f7b7f31866f7ad5bb1226 - [view on LGTM.com](https://lgtm.com/projects/g/cfengine/core/rev/pr-ed077bea6d23272c2a0b7925b7ec15ff4f666274)

**new alerts:**

* 1 for Wrong type of arguments to formatting function

---

*Comment posted by [LGTM.com](https://lgtm.com)*

olehermanse commented:
@cf-bottom jenkins with exotics, please.

cf-bottom commented:
Sure, I triggered a build:

[![Build Status](https://ci.cfengine.com//buildStatus/icon?job=pr-pipeline&build=1372)](https://ci.cfengine.com//job/pr-pipeline/1372/)

(with exotics)

https://ci.cfengine.com/job/pr-pipeline/1372/

olehermanse commented:
@cf-bottom jenkins with exotics, please.

cf-bottom commented:
Alright, I triggered a build:

[![Build Status](https://ci.cfengine.com//buildStatus/icon?job=pr-pipeline&build=1389)](https://ci.cfengine.com//job/pr-pipeline/1389/)

(with exotics)

https://ci.cfengine.com/job/pr-pipeline/1389/

olehermanse commented:
@cf-bottom let's also try a jenkins build with this new simplified StringCopy. With exotics!

cf-bottom commented:
Sure, I triggered a build:

[![Build Status](https://ci.cfengine.com//buildStatus/icon?job=pr-pipeline&build=1392)](https://ci.cfengine.com//job/pr-pipeline/1392/)

(with exotics)

https://ci.cfengine.com/job/pr-pipeline/1392/

olehermanse commented:
@cf-bottom jenkins with exotics

cf-bottom commented:
Alright, I triggered a build:

[![Build Status](https://ci.cfengine.com//buildStatus/icon?job=pr-pipeline&build=1398)](https://ci.cfengine.com//job/pr-pipeline/1398/)

(with exotics)

https://ci.cfengine.com/job/pr-pipeline/1398/

cf-bottom commented:
Sure, I triggered a build:

[![Build Status](https://ci.cfengine.com//buildStatus/icon?job=pr-pipeline&build=1399)](https://ci.cfengine.com//job/pr-pipeline/1399/)

(with exotics)

https://ci.cfengine.com/job/pr-pipeline/1399/

cf-bottom commented:
Sure, I triggered a build:

[![Build Status](https://ci.cfengine.com//buildStatus/icon?job=pr-pipeline&build=1402)](https://ci.cfengine.com//job/pr-pipeline/1402/)

(with exotics)

https://ci.cfengine.com/job/pr-pipeline/1402/

cf-bottom commented:
Sure, I triggered a build:

[![Build Status](https://ci.cfengine.com//buildStatus/icon?job=pr-pipeline&build=1405)](https://ci.cfengine.com//job/pr-pipeline/1405/)

(with exotics)

https://ci.cfengine.com/job/pr-pipeline/1405/

cf-bottom commented:
Alright, I triggered a build:

[![Build Status](https://ci.cfengine.com//buildStatus/icon?job=pr-pipeline&build=1406)](https://ci.cfengine.com//job/pr-pipeline/1406/)

(with exotics)

https://ci.cfengine.com/job/pr-pipeline/1406/

cf-bottom commented:
Alright, I triggered a build:

[![Build Status](https://ci.cfengine.com//buildStatus/icon?job=pr-pipeline&build=1416)](https://ci.cfengine.com//job/pr-pipeline/1416/)

(with exotics)

https://ci.cfengine.com/job/pr-pipeline/1416/

cf-bottom commented:
Sure, I triggered a build:

[![Build Status](https://ci.cfengine.com//buildStatus/icon?job=pr-pipeline&build=1417)](https://ci.cfengine.com//job/pr-pipeline/1417/)

(with exotics)

https://ci.cfengine.com/job/pr-pipeline/1417/

cf-bottom commented:
Alright, I triggered a build:

[![Build Status](https://ci.cfengine.com//buildStatus/icon?job=pr-pipeline&build=1418)](https://ci.cfengine.com//job/pr-pipeline/1418/)

(with exotics)

https://ci.cfengine.com/job/pr-pipeline/1418/

vpodzime commented on configure.ac:
```diff
@@ -994,6 +994,9 @@ AC_REPLACE_FUNCS(mkdtemp)
 
+AC_CHECK_DECLS(stpncpy)
+AC_REPLACE_FUNCS(stpncpy)
+
```
I would expect that something would have to be changed in *libcompat/Makefile.am* to add the new source file, but somehow this is enough, right?

olehermanse replied:
I would so too, but I think there's some magic going on with the filenames in that directory. I'm pretty sure it worked at the time I introduced it.

vpodzime commented on libutils/sequence.c:
```diff
@@ -449,3 +451,114 @@ int SeqStringLength(Seq *seq)
+        return -1;
+    }
+
+    // NOTE: This uses unsigned long because HPUX sscanf doesn't support %zu
```
Why not just use `long` here when it's type-casted to `long` at the end anyway?

olehermanse replied:
Thanks! I thought `%lu` wouldn't match negative numbers, but it does :(

vpodzime commented on libutils/sequence.c:
```diff
@@ -449,3 +451,114 @@ int SeqStringLength(Seq *seq)
+        if (length < 0
+            || NULL == (new_str = ValidDuplicate(src, length))
+            || src[-1] != ' '
+            || src[length] != '\n')
```
I know this version of the code is not focusing on performance, but still these checks could be reordered to do the most expensive one last.

vpodzime replied:
Oh, I see.

vpodzime commented on libutils/sequence.c:
```diff
@@ -449,3 +451,114 @@ int SeqStringLength(Seq *seq)
+    assert(serialized != NULL);
+    assert(SEQ_PREFIX_LEN > 0);
+
+    Seq *seq = SeqNew(512, free);
```
512 looks quite big to me. Is it based on some sample data? Even for reporting 512 CSV entries sounds quite overestimated.

olehermanse replied:
No more than intuition. Changed to `128` until I have checked with real data.

vpodzime commented on libutils/string_lib.c:
```diff
@@ -60,6 +60,72 @@ char *StringFormat(const char *fmt, ...)
+        }
+    }
+    return dst + len;
+}
```
Why is this needed here? It is the same as the implementation in *libcompat*, right?

olehermanse replied:
Yes, thanks, removed. I just added it while debugging.

vpodzime commented on libutils/string_lib.h:
```diff
@@ -190,6 +190,8 @@ char *StringVFormat(const char *fmt, va_list ap);
  */
 char *StringFormat(const char *fmt, ...) FUNC_ATTR_PRINTF(1, 2);
 
+long StringCopy(char *const dst, const char *const src, const long n);
```
Some of the `const`s here do not make much sense for a header file/prototype, but not a big deal.

vpodzime commented on tests/unit/sequence_test.c:
```diff
@@ -1,6 +1,6 @@
 #include <test.h>
 
-#include <sequence.h>
+#include <sequence.c>
```
Ouch... :laughing:

vpodzime replied:
Yeah, I know.

olehermanse commented:
@cf-bottom I have addressed the review comments, please trigger in jenkins with exotics! :)

cf-bottom commented:
Sure, I triggered a build:

[![Build Status](https://ci.cfengine.com//buildStatus/icon?job=pr-pipeline&build=1441)](https://ci.cfengine.com//job/pr-pipeline/1441/)

(with exotics)

https://ci.cfengine.com/job/pr-pipeline/1441/
